src/vm_pmu.rs

```diff
@@ -368,4 +373,149 @@ impl VmPmuState {
+            ) {
+                CSR.shpmevent.read_and_set_bits(event_type.raw());
+            } else {
+                CSR.shpmevent.read_and_set_bits(event_data);
+            }
+            CSR.shpmevent
+                .read_and_set_field(riscv_regs::defs::shpmevent::sinh);
+
+            if config_flags.is_clear_value() {
+                CSR.shpmcounter.set(0);
+            }
+
+            if config_flags.is_auto_start() {
+                CSR.scountinhibit
+                    .set(CSR.scountinhibit.read_and_clear_bits(1 << index));
+            }
+
+            Ok(index)
+        }
+    }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for start_counters() depending on whether Sscdeleg is implemented
+    pub fn start_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        start_flags: PmuCounterStartFlags,
+        initial_value: u64,
+    ) -> SbiResult<()> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::start_counters(counter_index, counter_mask, start_flags, initial_value)
+        } else {
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+            for index in bitmask_iter {
+                if matches!(self.counter_state[index], Configured(_)) {
+                    let index = index as u64;
+                    if start_flags.is_init_value() {
+                        CSR.scounterselect.set(index);
+                        CSR.shpmcounter.set(initial_value);
+                    }
+                    CSR.scountinhibit
+                        .set(CSR.scountinhibit.read_and_clear_bits(1 << index));
+                }
+            }
+            Ok(())
+        }
+    }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for stop_counters() depending on whether Sscdeleg is implemented
+    pub fn stop_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        stop_flags: PmuCounterStopFlags,
+    ) -> SbiResult<()> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::stop_counters(counter_index, counter_mask, stop_flags)
+        } else {
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+            for index in bitmask_iter {
+                if matches!(self.counter_state[index], Started(_)) {
+                    let index = index as u64;
+                    CSR.scountinhibit.read_and_set_bits(1 << index);
+                    if stop_flags.is_reset_flag() {
+                        CSR.scounterselect.set(index);
+                        CSR.shpmcounter.set(0);
```

Comment:
SBI_PMU_STOP_FLAG_RESET only resets the mapping doesn't reset the counter value.
shpmevent should be only reset.

Reply:
Ack...I had assumed that it reset the counter as well...will change.